File: 0238-product-of-array-except-self/0238-product-of-array-except-self.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int N = nums.length;
        int[] left_product = new int[N];
        int[] right_product = new int[N];
        int[] product = new int[N];
        int leftProduct = 1;
        int rightProduct = 1;

        for (int i = 0; i < N; i++) {
            if (i <= 0) {
                left_product[i] = 1;
            } else {
                leftProduct *= nums[i-1];
                left_product[i] = leftProduct;
            }
        }

        for (int i = N -1; i >= 0; i--) {
            if (i >= N -1) {
                right_product[i] = 1;
            } else {
                rightProduct *= nums[i+1];
                right_product[i] = rightProduct;
            }
            product[i] = left_product[i] * right_product[i];
        }
        return product;
    }
}

File: 0238-product-of-array-except-self/0238-product-of-array-except-self.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        res = [1] * len(nums)

        prefix = 1
        for i in range(len(nums)):
            res[i] *= prefix
            prefix *= nums[i]

        postfix = 1
        for i in range(len(nums) - 1, -1, -1):
            res[i] *= postfix
            postfix *= nums[i]
        
        return res

"""
Sol 1: O(n) space
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        left, right, res = [1] * len(nums), [1] * len(nums), [1] * len(nums)

        for i in range(1, len(nums)):
            left[i] = nums[i - 1] * left[i - 1]
        
        for i in range(len(nums) - 2, -1, -1):
            right[i] = right[i + 1] * nums[i + 1] 
        
        for i in range(len(nums)):
            res[i] = left[i] * right[i]
        
        return res
"""